src/thumb/addsub.rs

```diff
@@ -0,0 +1,123 @@
+use crate::RegisterOrImmediate;
+use crate::types::{Register, Immediate};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddSubtractOpCode {
+    ADD,
+    SUB
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct AddSubtract {
+    op: AddSubtractOpCode,
+    roi: RegisterOrImmediate,
+    src: Register,
+    dest: Register
+}
+
+impl From<u16> for AddSubtract {
+    fn from(raw: u16) -> Self { 
+        assert!((raw >> 11) == 0b00011);
+
+        AddSubtract {
+            op: AddSubtract::parse_opcode(raw),
+            roi: AddSubtract::parse_roi(raw),
+            src: Register(((raw >> 3) & 0b111) as u8),
+            dest: Register((raw & 0b111) as u8)
+        }
+    }
+}
+
+impl AddSubtract { 
+    fn parse_roi(raw: u16) -> RegisterOrImmediate { 
+        match (raw >> 10) & 1 {
+            0 => RegisterOrImmediate::Register(Register(((raw >> 6) & 0b111) as u8)),
+            1 => RegisterOrImmediate::Immediate(((raw >> 6) & 0b111) as u8),
+            _ => unreachable!()
+        } 
+    }
+
+    fn parse_opcode(raw: u16) -> AddSubtractOpCode {
+        let op_value = raw & (1 << 9);
```

Comment:
Consider doing `match (op_value >> 9) & 1` instead - for consistency and clarity